File: app/src/main/java/com/bapplications/maplemobile/opengl/utils/Rectangle.java
package com.bapplications.maplemobile.opengl.utils;

import com.bapplications.maplemobile.constatns.Configuration;

import com.bapplications.maplemobile.pkgnx.NXNode;
import com.bapplications.maplemobile.pkgnx.nodes.NXPointNode;

public class Rectangle {

    private Point left_top;
    private Point right_bottom;

    public Rectangle(NXPointNode sourceLeftTop, NXPointNode sourceRightBottom) {
        left_top = new Point(sourceLeftTop.getPoint());
        right_bottom = new Point(sourceRightBottom.getPoint());

        left_top.flipY();
        right_bottom.flipY();

    }

    public Rectangle(NXNode source){
        left_top = new Point(source.getChild("lt"));
        right_bottom = new Point(source.getChild("rb"));
        left_top.flipY();
        right_bottom.flipY();
    }

    public Rectangle(Point leftTop, Point rightBottom) {
        left_top = new Point(leftTop);
        right_bottom = new Point(rightBottom);
    }
    public Rectangle(float left, float right, float top, float bottom){
        left_top = new Point(left, top);
        right_bottom = new Point(right, bottom);
    }
    public Rectangle() {}

    public Rectangle(Rectangle rect) {
        left_top = new Point(rect.left_top);
        right_bottom = new Point(rect.right_bottom);
    }

    public float width()
    {
        return Math.abs(left() - right());
    }

    public float height()
    {
        return Math.abs(top() - bottom());
    }

    public float left() {
        return left_top.x;
    }

    public float top()
    {
        return left_top.y;
    }

    public float right()
    {
        return right_bottom.x;
    }

    public float bottom()
    {
        return right_bottom.y;
    }

    public boolean contains(Point p)
    {
        return !straight() &&
                        p.x >= left() && p.x <= right() &&
                        p.y <= top() && p.y >= bottom();
    }

    public boolean overlaps(Rectangle ar)
    {
            return get_horizontal().intersect(new Range<Float>(ar.left(), ar.right()))
                    && get_vertical().intersect(new Range<Float>(ar.top(), ar.bottom()));
    }

    public boolean straight()
    {
        return left_top == right_bottom;
    }

    public boolean empty()
    {
        return left_top.x == left_top.y && right_bottom.x == right_bottom.y && straight();
    }

    public Point get_left_top()
    {
        return left_top;
    }

    public Point get_right_bottom()
    {
        return right_bottom;
    }

    public Point get_right_top()
    {
        return new Point(right_bottom.x, left_top.y);
    }

    public Point get_left_bottom()
    {
        return new Point(left_top.x, right_bottom.y);
    }

    public Range<Float> get_horizontal()
    {
        return new Range<Float>( left(), right() );
    }

    public Range<Float> get_vertical()
    {
        return new Range<Float>(top(), bottom());
    }

    public void shift(Point v)
    {
        left_top.offset(v.x, v.y);
        right_bottom.offset(v.x, v.y);
    }

    public void draw(Point pos) {
        DrawArgument args = new DrawArgument(pos);
        DrawableCircle[] points = new DrawableCircle[4];
        points[0] = DrawableCircle.createCircle(get_left_top(), Color.GREEN);
        points[1] = DrawableCircle.createCircle(get_right_top(), Color.GREEN);
        points[2] = DrawableCircle.createCircle(get_left_bottom(), Color.GREEN);
        points[3] = DrawableCircle.createCircle(get_right_bottom(), Color.GREEN);
        for(DrawableCircle p: points) {
            p.draw(args);
        }
    }

}

File: app/src/main/java/com/bapplications/maplemobile/utils/Rectangle.kt
package com.bapplications.maplemobile.utils

import com.bapplications.maplemobile.pkgnx.NXNode
import com.bapplications.maplemobile.pkgnx.nodes.NXBitmapNode
import com.bapplications.maplemobile.pkgnx.nodes.NXPointNode
import com.bapplications.maplemobile.utils.Point.TwoDPolygon


class Rectangle(var leftTop: Point, var rightBottom: Point) : TwoDPolygon {

    // want new instance of the points to avoid shallow copy
    init {
        leftTop = Point(leftTop)
        rightBottom = Point(rightBottom)
    }
    
    constructor(sourceLeftTop: NXPointNode, sourceRightBottom: NXPointNode) :
            this(Point(sourceLeftTop.point).flipY(), Point(sourceRightBottom.point).flipY())

    constructor(source: NXNode) :
            this(Point(source.getChild("lt")).flipY(),
                    Point(source.getChild("rb")).flipY()) {
        if(source.getChild<NXNode>("lt").isNotExist
                && source is NXBitmapNode) {
            val bmap = source.get()
            this.leftTop = Point(-bmap.width/2, bmap.height)
            this.rightBottom = Point(bmap.width/2, 0)
            bmap.recycle()
        }
    }

    constructor(horizon: Range, vertical: Range): this(Point(horizon.lower, vertical.upper), Point(horizon.upper, vertical.lower))
    
    constructor(left: Float, right: Float, top: Float, bottom: Float): this(Point(left, top), Point(right, bottom)) 

    constructor() : this(Point(), Point())

    constructor(rect: Rectangle): this(Point(rect.leftTop), Point(rect.rightBottom))

    fun width(): Float {
        return Math.abs(left() - right())
    }

    fun height(): Float {
        return Math.abs(top() - bottom())
    }

    fun left(): Float {
        return leftTop.x
    }

    fun top(): Float {
        return leftTop.y
    }

    fun right(): Float {
        return rightBottom.x
    }

    fun bottom(): Float {
        return rightBottom.y
    }

    fun center(): Point {
        return Point((left() + right()) / 2, (top() + bottom()) / 2)
    }

    operator fun contains(p: Point): Boolean {
        return !straight() && p.x >= left() && p.x <= right() && p.y <= top() && p.y >= bottom()
    }

    fun overlaps(ar: Rectangle): Boolean {
        return (get_horizontal().intersect(Range(ar.left(), ar.right()))
                && get_vertical().intersect(Range(ar.top(), ar.bottom())))
    }

    fun straight(): Boolean {
        return leftTop === rightBottom
    }

    fun empty(): Boolean {
        return leftTop.x == leftTop.y && rightBottom.x == rightBottom.y && straight()
    }

    fun get_left_top(): Point? {
        return leftTop
    }

    fun get_right_bottom(): Point? {
        return rightBottom
    }

    fun get_right_top(): Point {
        return Point(rightBottom.x, leftTop.y)
    }

    fun get_left_bottom(): Point {
        return Point(leftTop.x, rightBottom.y)
    }

    fun get_horizontal(): Range {
        return Range(left(), right())
    }

    fun get_vertical(): Range {
        return Range(top(), bottom())
    }

    fun shift(v: Point) {
        leftTop.offset(v.x, v.y)
        rightBottom.offset(v.x, v.y)
    }

    fun setLeft(`val`: Float) {
        leftTop.x = `val`
    }

    fun setRight(`val`: Float) {
        rightBottom.x = `val`
    }

    fun draw(pos: Point) {
        val args = DrawArgument(pos)
        val points = listOf<DrawableCircle>(
                DrawableCircle.createCircle(get_left_top(), Color.GREEN),
                DrawableCircle.createCircle(get_right_top(), Color.GREEN),
                DrawableCircle.createCircle(get_left_bottom(), Color.GREEN),
                DrawableCircle.createCircle(get_right_bottom(), Color.GREEN))
        for (p in points) {
            p.draw(args)
        }
    }

    override fun toString(): String {
        return "$leftTop -> $rightBottom"
    }

    override val width: Range
        get() = get_horizontal()
    override val height: Range
        get() = get_vertical()
}